### PR TITLE
Add GLEM model, TAGDataset & running example

### PR DESCRIPTION
**Pull request by ECMGit**
Feature summary:

- Add GLEM as GNN & LLM Co-training model to PyG
- adapt GLEM's LM to AutoModelForSequenceClassification from transformers
- Lora support
- LM/LLM support
- ogbn-products/ogbn-arxiv testing finished
- TAGDataset can be used as a wrapper class for any node classification dataset in PyG with LM tokenizer and associate raw text
- external prediction as pseudo labels supported

### CODE REVIEW

File: examples/llm_plus_gnn/README.md [hunk 1]
```diff
@@ -1,5 +1,6 @@
...
-| [`g_retriever.py`](./g_retriever.py) | Example for Retrieval-Augmented Generation (RAG) w/ GNN+LLM by co-training LLAMA2 with GAT for answering questions based on knowledge graph information |
+| Example                              | Description                                                                                                                                                            |
+| ------------------------------------ | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| [`g_retriever.py`](./g_retriever.py) | Example for Retrieval-Augmented Generation (RAG) w/ GNN+LLM by co-training LLAMA2 with GAT for answering questions based on knowledge graph information                |
+| [`glem.py`](./glem.py)               | Example for [GLEM](https://arxiv.org/abs/2210.14709), a GNN+LLM co-training model via variational Expectation-Maximization (EM) framework on node classification tasks |
```
Comment (puririshi98) [inline examples/llm_plus_gnn/README.md:6]: add 
"framework on node classification tasks to achieve SOTA results"

Comment (puririshi98) [inline examples/llm_plus_gnn/README.md:6]: add
"framework on node classification tasks to achieve SOTA results"

**Review by puririshi98 — CHANGES_REQUESTED** [review]

File: examples/llm_plus_gnn/glem.py [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+                                  drop_last=False, pin_memory=False,
+                                  shuffle=False)
+    print('done')
+
+    # =========================== GNN Data Loader =============================
+    initial_memory = torch.cuda.memory_allocated()
+    data = data.to(device)
+
+    current_memory_1 = torch.cuda.max_memory_allocated()
+    gpu_usage = float(current_memory_1 - initial_memory) / 1073741824
```
Comment (puririshi98) [inline examples/llm_plus_gnn/glem.py:129]: put comment explaining what the hardcoded value means

**Review by puririshi98 — CHANGES_REQUESTED** [review]

File: examples/llm_plus_gnn/glem.py [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+    parser.add_argument('--beta', type=float, default=0.5,
+                        help='pseudo label weight in M-step')
+    parser.add_argument('--lm_epochs', type=int, default=10)
+    parser.add_argument('--gnn_epochs', type=int, default=50)
+    parser.add_argument('--gnn_lr', type=float, default=0.003)
+    parser.add_argument('--lm_lr', type=float, default=0.001)
+    parser.add_argument('--patience', type=int, default=5)
+    parser.add_argument('--verbose', action='store_true',
+                        help='show progress bar during training or not')
+    parser.add_argument('--em_order', type=str, default='lm')
```
Comment (puririshi98) [inline examples/llm_plus_gnn/glem.py:364]: add help to explain that this decides wether to train LM first or GNN first

**Review by puririshi98 — CHANGES_REQUESTED** [review]

File: examples/llm_plus_gnn/glem.py [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+    print(tag_dataset.num_classes, tag_dataset.raw_file_names)
+
+    num_classes = tag_dataset.num_classes
+    num_features = data.num_features
+    # =========================== LM Data split ===============================
+    split_idx = tag_dataset.get_idx_split()
+
+    # GLEM train with augmented data, mark original train data as gold data,
+    gold_idx = split_idx['train']
+    split_idx['valid']
```
Comment (puririshi98) [inline examples/llm_plus_gnn/glem.py:85]: make sure to assign to a variable

File: examples/llm_plus_gnn/glem.py [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+    train_dataset = torch.utils.data.Subset(dataset=text_dataset,
+                                            indices=train_idx)
+    # valid_dataset = torch.utils.data.Subset(dataset=text_dataset,
+    #                                         indices=valid_idx)
+
+    # ========================== LM Data Loader ===============================
+
+    print('Building language model dataloader...', end='-->')
+    from torch_geometric.loader import DataLoader
+    text_pretrain_loader = DataLoader(gold_dataset, batch_size=lm_batch_size,
```
Comment (puririshi98) [inline examples/llm_plus_gnn/glem.py:110]: would be good to check what happens if you set pin_memory=True, if makes faster, would be worth turning it on

**Review by puririshi98 — CHANGES_REQUESTED** [review]

File: examples/llm_plus_gnn/glem.py [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+                        help="pseudo labels ratio")
+    parser.add_argument('--hf_model', type=str, default='prajjwal1/bert-tiny',
+                        help='huggingface model repo id')
+    parser.add_argument(
+        '--gnn_model', type=str, default='SAGE',
+        help='gnn model for node classification,'
+        'options: SAGE, GAT, GCN')
+    parser.add_argument('--lm_batch_size', type=int, default=256)
+    parser.add_argument('--gnn_batch_size', type=int, default=1024)
+    parser.add_argument('--external_pred_path', type=str, default=None)
```
Comment (puririshi98) [inline examples/llm_plus_gnn/glem.py:352]: add descriptive help string to explain that these external predictions will be used for label augmentation

**Review by puririshi98 — CHANGES_REQUESTED** [review]

File: examples/llm_plus_gnn/glem.py [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+            test_acc = evaluator.eval({
+                'y_true': y_true[split_idx['test']],
+                'y_pred': y_pred[split_idx['test']],
+            })['acc']
+
+        return train_acc, val_acc, test_acc
+
+    # =========================== Build GNN Model =============================
+    gnn = None
+    if args.gnn_model == 'SAGE':
```
Comment (puririshi98) [inline examples/llm_plus_gnn/glem.py:196]: this should be controlled by the argparser (hidden channels, num layers etc, make those defaults)

**Review by puririshi98 — CHANGES_REQUESTED** [review]

File: examples/llm_plus_gnn/glem.py [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+
+    # graph data loader w/ pseudo labels in M-step
+    graph_train_loader = NeighborLoader(
+        data,
+        input_nodes=train_idx,
+        num_neighbors=[15, 10, 5],
+        batch_size=gnn_batch_size,
+        shuffle=True,
+        num_workers=12,
+        persistent_workers=True,
```
Comment (puririshi98) [inline examples/llm_plus_gnn/glem.py:155]: im curious what happens if you set persistent_workers=False, please test it and report the outcome (change in speed etc) so that we can decide the best default

File: examples/llm_plus_gnn/glem.py [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+    parser.add_argument('--gnn_epochs', type=int, default=50)
+    parser.add_argument('--gnn_lr', type=float, default=0.003)
+    parser.add_argument('--lm_lr', type=float, default=0.001)
+    parser.add_argument('--patience', type=int, default=5)
+    parser.add_argument('--verbose', action='store_true',
+                        help='show progress bar during training or not')
+    parser.add_argument('--em_order', type=str, default='lm')
+    parser.add_argument('--lm_use_lora', action='store_true',
+                        help='use Lora to fine-tune model or not')
+    parser.add_argument('--token_on_disk', action='store_true')
```
Comment (puririshi98) [inline examples/llm_plus_gnn/glem.py:367]: add help to explain that this is used to save the tokens on disk in case RAM is insufficient

File: examples/llm_plus_gnn/glem.py [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+    parser.add_argument('--external_pred_path', type=str, default=None)
+    parser.add_argument('--alpha', type=float, default=0.5,
+                        help='pseudo label weight in E-step')
+    parser.add_argument('--beta', type=float, default=0.5,
+                        help='pseudo label weight in M-step')
+    parser.add_argument('--lm_epochs', type=int, default=10)
+    parser.add_argument('--gnn_epochs', type=int, default=50)
+    parser.add_argument('--gnn_lr', type=float, default=0.003)
+    parser.add_argument('--lm_lr', type=float, default=0.001)
+    parser.add_argument('--patience', type=int, default=5)
```
Comment (puririshi98) [inline examples/llm_plus_gnn/glem.py:361]: help="Patience for early stopping"

**Review by puririshi98 — CHANGES_REQUESTED** [review]

File: examples/llm_plus_gnn/glem.py (hunk 2 above)
Comment (ECMGit) [reply examples/llm_plus_gnn/glem.py:129]: 1 GB = 1073741824 Bytes

File: examples/llm_plus_gnn/glem.py (hunk 5 above)
Comment (ECMGit) [reply examples/llm_plus_gnn/glem.py:110]: pin_memory is a argument for `torch.utils.data.DataLoader` not include in `torch_geometric.loader.DataLoader` I have remove it in current example

Comment (ECMGit) [reply examples/llm_plus_gnn/glem.py:110]: actually, It does not make any difference for speed after testing

File: examples/llm_plus_gnn/glem.py (hunk 8 above)
Comment (ECMGit) [reply examples/llm_plus_gnn/glem.py:155]: test gnn training 50 epochs w/ full inference on ogbn-arxiv, running time 32 seconds when `persistent_workers=True` --> 85 seconds when `persistent_workers=False`, memory usage not changing much

Comment (puririshi98) [reply examples/llm_plus_gnn/glem.py:155]: okay good, lets keep True for now

File: examples/llm_plus_gnn/glem.py (hunk 5 above)
Comment (puririshi98) [reply examples/llm_plus_gnn/glem.py:110]: okay lets leave it removed then

File: examples/llm_plus_gnn/glem.py (hunk 2 above)
Comment (puririshi98) [reply examples/llm_plus_gnn/glem.py:129]: yes a comment like "# GPU usage, in GB" would be good

File: examples/llm_plus_gnn/glem.py [hunk 11]
```diff
@@ -0,0 +1,393 @@
...
+
+    print('Building language model dataloader...', end='-->')
+    from torch_geometric.loader import DataLoader
+    text_pretrain_loader = DataLoader(gold_dataset, batch_size=lm_batch_size,
+                                      drop_last=False, pin_memory=True,
+                                      shuffle=True)
+    text_train_loader = DataLoader(train_dataset, batch_size=lm_batch_size,
+                                   drop_last=False, pin_memory=True,
+                                   shuffle=True)
+    # text_val_loader = DataLoader(valid_dataset, batch_size=lm_batch_size * 4,
```
[the remark this answers is not in the document]
Comment (ECMGit) [reply examples/llm_plus_gnn/glem.py]: done

**Comment by puririshi98** [comment]
LGTM

**Comment by puririshi98** [comment]
@ECMGit  please reopen by making this merge into master since my branch was merged and deleted
